### DISCUSSION TITLE
Correct way to access request domain cookie on nextjs server side & pass to graphql server (resolved)

### DISCUSSION THREAD

**Discussion by Nosherwan** [urql-graphql/urql#2568] [Q&A] [answered]
urql version 2.2.3
next js version 12.2.2
next-urql version 3.3.3

I am using urql to fetch data from an apollo GraphQL server via SSR.
I am using WithUrqlClient for this purpose.
I also pass the request header cookie from the requesting browser to the client as the authentication is done via httpOnly cookie on the website & is the only way to pass secure cookie to the server. (i suspect this is not related to the issue)

The page works correctly, however if I reload the page in the browser via reload button; it throws the errors mentioned below, only when I try to access the "error" variable from the "result" object returned by the useQuery hook.

If I comment out the error the page works as intented and no errors are thrown.
Also when the user is redirected from another page to this one the errors are not thrown for the first server access either.

Errors I get:
Error: Text content does not match server-rendered HTML.
Error: There was an error while hydrating. Because the error happened outside of a Suspense boundary, the entire root will switch to client rendering.
 
Below I am providing the minimal code, that can be executed.

```ts
import Head from 'next/head';
import { useQuery } from 'urql';
import { withUrqlClient } from 'next-urql';
import { url } from '../graphql/client';
import { UserQuery } from '../graphql/operations/user.gql';
import { useState } from 'react';

const UsersPage = ({
	pageSize,
	after,
	filter,
	isLastPage,
	onLoadMore,
}: {
	pageSize: number;
	after: number;
	filter: string[];
	isLastPage: boolean;
	onLoadMore: any;
}) => {
	const [result, reexecuteQuery] = useQuery({
		query: UserQuery,
		variables: { pageSize, after, filter },
		// pause: true,
	});
	const { data, fetching, error } = result;
	const users = data?.users?.users ?? [];
	const more = data?.users?.hasMore ?? false;
	const cursor = data?.users?.cursor ?? 0;

	return (
		<div>
			{/* ERROR: uncommenting the below line gives the error */}
			{error && <p>Oh no...{error.message}</p>}
			{fetching && <p>Loading...</p>}
			{users.length > 0 && (
				<>
					{users.map((user: any) => (
						<div key={user.email}>{user.email}</div>
					))}
				</>
			)}
			{isLastPage && more && (
				<button onClick={() => onLoadMore(+cursor)}>Load More</button>
			)}
		</div>
	);
};

const Home = () => {
	const pageSize = 2;

	const [pageVars, setPageVars] = useState([
		{
			pageSize,
			after: 0,
			filter: [],
		},
	]);

	return (
		<>
			<Head>
				<title>Page</title>
			</Head>
			<main>
				{pageVars.map((vars: any, i: number) => (
					<UsersPage
						key={vars.after}
						pageSize={vars.pageSize}
						after={vars.after}
						filter={vars.filter}
						isLastPage={i === pageVars.length - 1}
						onLoadMore={(after: number) =>
							setPageVars([...pageVars, { pageSize, after, filter: [] }])
						}
					/>
				))}
			</main>
		</>
	);
};

export default withUrqlClient(
	(_ssrExchange, ctx: any) => {
		const cookie = ctx?.req?.cookies ?? '';

		return {
			url,
			fetchOptions: () => ({
				credentials: 'include',
				headers: { cookie },
			}),
		};
	},
	{ ssr: true }
)(Home);

```

**Comment by JoviDeCroock** [comment, marked as answer]
You will need to look at what the difference is on the client vs server, it could be that your request is erroring out on either side so your best bet it to go to your network tab and look at the HTML your server has sent back to you and then compare it to what you are seeing in your client DOM.

Next up would be to see what query errors out/.... generally this error occurs when the server for instance sends a loading state and your client works further on that.

The information from your code-fragments is generally insufficient for us to exactly tell you what is going on because we are missing these two states (client and server), hence your best bet is to inspect  the difference and see _why_ your server/client interaction is showing this difference.

My assumption is that your server has an `error` and `data` at the same time (which is possible with GraphQL) hence you are seeing this in the start. All though errors should (de)serialize correctly https://github.com/FormidableLabs/urql/blob/main/packages/core/src/exchanges/ssr.ts#L69

**Reply by Nosherwan** [reply]
Thanks @JoviDeCroock you pointed me in the right direction, just so that someone else can figure out if they have the same issue here is how I resolved it.

So the server side request (SSR) was the one that was erroring out and because of that next js was switching to client side rendering.
That gave me the clue that the way I was assigning the cookie was incorrect.

So I was using the following code to send the cookie back from next.js to the apollo GraphQL server.

```ts
export default withUrqlClient(
	(_ssrExchange, ctx: any) => {
                // The line below is incorrect
		const cookie = ctx?.req?.cookies ?? '';

		return {
			url,
			fetchOptions: () => ({
				credentials: 'include',
				headers: { cookie },
			}),
		};
	},
	{ ssr: true }
)(Home);
```
So I had to extract the correct cookie value from the header and not the req.cookies, as below:

```ts
export default withUrqlClient(
	(_ssrExchange, ctx: any) => {
		//This is the correct way of fetching the cookie
		const cookie = ctx?.req?.headers?.cookie ?? '';

		return {
			url,
			fetchOptions: () => ({
				credentials: 'include',
				headers: { cookie },
			}),
		};
	},
	{ ssr: true }
)(Home);
```
And that fixed the error.
